test/python/circuit/test_equivalence.py

```diff
@@ -0,0 +1,346 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+
+"""Test Qiskit's EquivalenceLibrary class."""
+
+import numpy as np
+
+from qiskit.test import QiskitTestCase
+
+from qiskit.circuit import QuantumCircuit, Parameter, Gate
+from qiskit.circuit.exceptions import CircuitError
+
+from qiskit.circuit import EquivalenceLibrary
+
+
+class OneQubitZeroParamGate(Gate):
+    """Mock one qubit zero param gate."""
+    def __init__(self):
+        super().__init__('1q0p', 1, [])
+
+
+class OneQubitOneParamGate(Gate):
+    """Mock one qubit one  param gate."""
+    def __init__(self, theta):
+        super().__init__('1q1p', 1, [theta])
+
+
+class OneQubitTwoParamGate(Gate):
+    """Mock one qubit two param gate."""
+    def __init__(self, phi, lam):
+        super().__init__('1q2p', 1, [phi, lam])
+
+
+class TestEquivalenceLibraryWithoutBase(QiskitTestCase):
+    """Test cases for basic EquivalenceLibrary."""
+
+    def test_create_empty_library(self):
+        """An empty library should return an empty entry."""
+        eq_lib = EquivalenceLibrary()
+        self.assertIsInstance(eq_lib, EquivalenceLibrary)
+
+        gate = OneQubitZeroParamGate()
+
+        entry = eq_lib.get_entry(gate)
+
+        self.assertEqual(len(entry), 0)
+
+    def test_add_single_entry(self):
+        """Verify an equivalence added to the library can be retrieved."""
+        eq_lib = EquivalenceLibrary()
+
+        gate = OneQubitZeroParamGate()
+        equiv = QuantumCircuit(1)
+        equiv.h(0)
+
+        eq_lib.add_equivalence(gate, equiv)
+
+        entry = eq_lib.get_entry(gate)
+
+        self.assertEqual(len(entry), 1)
+        self.assertIsNot(entry[0], equiv)
+        self.assertEqual(entry[0], equiv)
+
+    def test_add_double_entry(self):
+        """Verify separately added equivalences can be retrieved."""
+        eq_lib = EquivalenceLibrary()
+
+        gate = OneQubitZeroParamGate()
+        first_equiv = QuantumCircuit(1)
+        first_equiv.h(0)
+
+        eq_lib.add_equivalence(gate, first_equiv)
+
+        second_equiv = QuantumCircuit(1)
+        second_equiv.u2(0, np.pi, 0)
+
+        eq_lib.add_equivalence(gate, second_equiv)
+
+        entry = eq_lib.get_entry(gate)
+
+        self.assertEqual(len(entry), 2)
+        self.assertTrue(
+            (entry[0] == first_equiv and entry[1] == second_equiv)
+            or (entry[1] == first_equiv and entry[0] == second_equiv))
```

Comment:
maybe compare sets? 
```suggestion
        self.assertEqual(set(entry), set([first_equiv, second_equiv])
```
Why not return them in order?

Reply:
Was on the fence about what the ordering should be. Ideas? (Insertion order oldest to newest from self to base is what we'll get now, I think.)

Reply:
my original comments was for being independent of the order. But I see value in keeping some deterministic order. That might be a proto-way to define priorities. I think `entry[0] == newest`... maybe a stack?

Reply:
I would've preferred a set as well (for the return type and the assertion) but `QuantumCircuit`s aren't hashable (as they're mutable). Agree deterministic order is good, but the caller should inspect the returned circuit to determine priority, so the exact order shouldn't be so important. I'll document the current order and remove the `or`s in the test.

Reply:
Done in 7378cec .